### PR TITLE
[wallet]Store encrypted local signing key

### LINKED ISSUE #1187: Users SBAT to store encrypted key for local signing
### Expected Behavior

Users SBAT to store encrypted key for local signing

### Current Behavior

Currently, the key for local signing is stored in the plain-text on the private storage.

### PR DESCRIPTION
**Pull request by ashishb**
### Description

Store encrypted local signing key, the encryption password is coming
from the Android Keystore.

#### Unit test

`$ node_modules/.bin/jest src/web3/saga.test.ts`

#### Manual test

1. Set `DEFAULT_SYNC_MODE=-1` in `packages/mobile/.env`
2. `packages/mobile $ yarn run dev`. Log in verify that you can send money.
3. Check the contents of the file `/data/data/$APP_ID/files/private_key_for_<address>.txt` are encrypted hex-encoded key `
8672a59c5d55302a5d64288084137303f5cd431abbff3d1edeca4b1c9f8266c41148d712c99d26b52165a6426b048ab1710c6311154c66e75e071386538c7a0091a920d1167d729a8916ad646099adfd`

#### 

### Related issues

- Fixes #1187

### Backwards compatibility

Not backward compatible but this code path was added yesterday (Oct 1) and is still behind a flag. So, it is not an issue.

### CODE REVIEW

File: packages/mobile/src/web3/saga.ts [hunk 1]
```diff
@@ -230,8 +235,34 @@ export async function readPrivateKeyFromLocalDisk(
...
+  return privateKey
+}
+
+// Exported for testing
+export function getEncryptedData(plainTextData: string, password: string): Buffer {
+  try {
+    const cipher = Crypto.createCipher('aes-256-cbc', password)
+    return Buffer.concat([cipher.update(new Buffer(plainTextData, 'utf8')), cipher.final()])
+  } catch (e) {
+    Logger.error('getEncryptedData', 'Failed to write private key')
```
Comment (jmrossy) [inline packages/mobile/src/web3/saga.ts]: use `TAG + ` in these error messages and include e as the 3rd param

File: packages/mobile/src/web3/saga.ts [hunk 2]
```diff
@@ -230,8 +235,34 @@ export async function readPrivateKeyFromLocalDisk(
...
+}
+
+// Exported for testing
+export function getDecryptedData(encryptedData: Buffer, password: string): string {
+  try {
+    const decipher = Crypto.createDecipher('aes-256-cbc', password)
+    const decrypted = Buffer.concat([decipher.update(encryptedData), decipher.final()])
+    return decrypted.toString('utf8')
+  } catch (e) {
+    Logger.error('getDecryptedData', 'Failed to read private key')
```
Comment (jmrossy) [inline packages/mobile/src/web3/saga.ts]: ditto

File: packages/mobile/src/web3/saga.ts [hunk 3]
```diff
@@ -203,6 +204,9 @@ function getPrivateKeyFilePath(account: string): string {
 
 function ensureAddressAndKeyMatch(address: string, privateKey: string) {
   const generatedAddress = getAccountAddressFromPrivateKey(privateKey)
+  if (!generatedAddress) {
+    throw new Error(`Failed to generate address from private key`)
+  }
```
Comment (jmrossy) [inline packages/mobile/src/web3/saga.ts:209]: not sure how a `console.debug` slipped in below (215) but please convert it to Logger

File: packages/mobile/src/web3/saga.ts [hunk 4]
```diff
@@ -230,8 +235,34 @@ export async function readPrivateKeyFromLocalDisk(
 ): Promise<string> {
   const filePath = getPrivateKeyFilePath(account)
   Logger.debug('readPrivateKeyFromLocalDisk', `Reading private key from ${filePath}`)
-  // TODO(ashishb): Read and decrypt private key instead
-  return RNFS.readFile(getPrivateKeyFilePath(account))
+  const hexEncodedEncryptedData: string = await RNFS.readFile(filePath)
```
Comment (jmrossy) [inline packages/mobile/src/web3/saga.ts:238]: Generally, these explicit typings (`varName: type`) aren't necessary unless the function/object your calling is missing a type, which usually shouldnt be the case

File: packages/mobile/src/web3/saga.ts [hunk 5]
```diff
@@ -218,9 +222,10 @@ async function savePrivateKeyToLocalDisk(
 ) {
   ensureAddressAndKeyMatch(account, privateKey)
   const filePath = getPrivateKeyFilePath(account)
-  Logger.debug('savePrivateKeyToLocalDisk', `Writing private key to ${filePath}`)
-  // TODO(ashishb): Store encrypted private key instead
-  await RNFS.writeFile(getPrivateKeyFilePath(account), privateKey)
+  const plainTextData = privateKey
```
Comment (annakaz) [inline packages/mobile/src/web3/saga.ts:225]: nit: why not skip the variable assignment and just pass privateKey into the function

File: packages/mobile/src/web3/saga.ts [hunk 6]
```diff
@@ -230,8 +235,34 @@ export async function readPrivateKeyFromLocalDisk(
 ): Promise<string> {
   const filePath = getPrivateKeyFilePath(account)
   Logger.debug('readPrivateKeyFromLocalDisk', `Reading private key from ${filePath}`)
-  // TODO(ashishb): Read and decrypt private key instead
-  return RNFS.readFile(getPrivateKeyFilePath(account))
+  const hexEncodedEncryptedData: string = await RNFS.readFile(filePath)
+  const encryptedDataBuffer: Buffer = new Buffer(hexEncodedEncryptedData, 'hex')
```
Comment (annakaz) [inline packages/mobile/src/web3/saga.ts:239]: @jmrossy On the other hand, I find this explicit typing to help with readability in cases that switch between strings/buffers/etc
Some linting rules will prevent explicit typing for common types like numbers, strings (ex: https://palantir.github.io/tslint/rules/no-inferrable-types/) but in the case of less common types I find the explicit typing helpful

**Review by annakaz — APPROVED** [review]
LGTM 👍

File: packages/mobile/src/web3/saga.ts (hunk 1 above)
Comment (ashishb) [reply packages/mobile/src/web3/saga.ts]: Done.

File: packages/mobile/src/web3/saga.ts (hunk 3 above)
Comment (ashishb) [reply packages/mobile/src/web3/saga.ts:209]: Done

File: packages/mobile/src/web3/saga.ts (hunk 4 above)
Comment (ashishb) [reply packages/mobile/src/web3/saga.ts:238]: It is not necessary, but I find it helpful when there is a mix going on.

```
const hexEncodedEncryptedData: string = await RNFS.readFile(filePath)
const encryptedDataBuffer: Buffer = new Buffer(hexEncodedEncryptedData, 'hex')
const privateKey: string = getDecryptedData(encryptedDataBuffer, encryptionPassword)
```

File: packages/mobile/src/web3/saga.ts (hunk 5 above)
Comment (ashishb) [reply packages/mobile/src/web3/saga.ts:225]: `getEncryptedData(privateKey, encryptionPassword)` felt a bit confusing to read.

File: packages/mobile/src/web3/saga.ts (hunk 3 above)
Comment (ashishb) [reply packages/mobile/src/web3/saga.ts:209]: Removed.

File: packages/mobile/src/web3/saga.ts (hunk 2 above)
Comment (ashishb) [reply packages/mobile/src/web3/saga.ts]: Fixed.

**Review by jmrossy — APPROVED** [review]

**Pull request merged**